Mathlib/Tactic/DeriveFintype.lean

```diff
@@ -0,0 +1,386 @@
+/-
+Copyright (c) 2023 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import Lean
+import Mathlib.Data.Fintype.Basic
+import Mathlib.Data.Fintype.Sigma
+import Mathlib.Data.Fintype.Sum
+
+/-!
+# The `Fintype` derive handler
+
+This file defines a derive handler to automatically generate `Fintype` instances
+for structures and inductive types.
+
+The following is a prototypical example of what this can handle:
+```
+inductive MyOption (α : Type _)
+  | none
+  | some (x : α)
+  deriving Fintype
+```
+
+This deriving handler does not attempt to process inductive types that are either
+recursive or that have indices.
+
+To get debugging information, do `set_option trace.Elab.Deriving.fintype true`.
+
+There is a term elaborator `derive_fintype%` implementing the derivation of `Fintype` instances.
+This can be useful in cases when there are necessary additional assumptions (like `DecidableEq`).
+
+Underlying this is a term elaborator `proxy_equiv%` for creating an equivalence from a
+"proxy type" composed of basic type constructors to the inductive type.
+
+## Implementation notes
+
+There are two kinds of `Fintype` instances that we generate, depending on the inductive type.
+
+If it is an enum (an inductive type with only 0-ary constructors), then we generate the
+complete `List` of all constructors; see `Mathlib.Deriving.Fintype.mkFintypeEnum` for more
+details. The proof has $O(n)$ complexity in the number of constructors.
+
+Otherwise, the strategy we take is to generate a "proxy type", define an equivalence between
+our type and the proxy type (see `proxy_equiv%`), and then use `Fintype.ofEquiv` to pull a
+`Fintype` instance on the proxy type (if one exists) to our inductive type. For example, with
+the `MyOption α` type above, we generate `Unit ⊕ α`. While the proxy type is not a finite type
+in general, we add `Fintype` instances for every type parameter of our inductive type (and
+`Decidable` instances for every `Prop` parameter). Hence, in our example we get
+`Fintype (MyOption α)` assuming `Fintype α`.
+
+There is a source of quadratic complexity in this `Fintype` instance from the fact that an
+inductive type with `n` constructors has a proxy type of the form `C₁ ⊕ (C₂ ⊕ (⋯ ⊕ Cₙ))`,
+so mapping to and from `Cᵢ` requires looking through `i` levels of `Sum` constructors.
+Ignoring time spent looking through these constructors, the construction of `Finset.univ`
+contributes linear time with respect to the cardinality of the type since the instances
+involved compute the underlying `List` for the `Finset` as `l₁ ++ (l₂ ++ (⋯ ++ lₙ))` with
```

Comment:
What makes `l₁ ++ (l₂ ++ (⋯ ++ lₙ))` quadratic complexity? Isn't it $O(mn)$, where $m$ is the number of constructors and $n$ is the size of each? So for the enum case, $n=1$, and the complexity should be the same.

Reply:
[`List.append`](https://leanprover-community.github.io/mathlib4_docs/Init/Data/List/Basic.html#List.append) agrees that `xs ++ ys` is $O(\mathtt{xs.length})$$

Reply:
Right, the point is that `l₁ ++ (l₂ ++ (⋯ ++ lₙ))` takes linear time in the total length of the resulting list. `((l₁ ++ l₂) ++ ⋯ ) ++ lₙ` would be quadratic time, which is why it's associated the way it is.

But, there's quadratic time (with respect to `n`) that's still there from needing to deal with all the `Sum` constructors. `Fintype.ofEquiv` maps over the resulting list.